### PR TITLE
Replace Hashie::Mash with ActiveSupport::HashWithIndifferentAccess

### PR DESCRIPTION
**Pull request by msroot**
This will replace `Hashie::Mash` with `ActiveSupport::HashWithIndifferentAccess`

### CODE REVIEW

File: spec/grape/validations/validators/coerce_spec.rb [hunk 1]
```diff
@@ -285,7 +285,7 @@ class User
           requires :file, type: Rack::Multipart::UploadedFile
         end
         subject.post '/upload' do
-          params[:file].filename
+          params[:file][:filename]
```
Comment (dblock) [inline spec/grape/validations/validators/coerce_spec.rb]: So this is a change in API as only the top keys are with indifferent access?

**Comment by dblock** [comment]
This is similar to #1448, but better.

I think we also need to provide a backward compatible way for existing users to return a `Hashie::Mash`. Here's what I would like the DSL to be:

```ruby
class API < Grape::API
  # params is a regular Hash
end
```

```ruby
class API < Grape::API
   include Grape::Extensions::...::Hashie::Mash
   # params is Hashie::Mash
end
```

```ruby
class API < Grape::API
   include Grape::Extensions::...::ActiveSupport::HashWithIndifferentAccess
   # params is HashWithIndifferentAccess
end
```

You get the idea ...

File: spec/grape/validations/validators/coerce_spec.rb (hunk 1 above)
Comment (msroot) [inline spec/grape/validations/validators/coerce_spec.rb]: nope! to all sub keys

**Comment by msroot** [comment]
@dblock sure but the default should be `ActiveSupport::HashWithIndifferentAccess` yes?
For the  `Hashie::Mash` does the user has to explicitly require the Hashie::Mash gem then?

About the DSL, is this something like this?

```
if defined? Hashie::Mash
params = Hashie::Mash.new(rack_params)
```

**Comment by dblock** [comment]
Yes, I think the default should be `Hash`.

The user should explicitly include `hashie` in their Gemfile in order to use the extension, and we should not have it in `.gemspec` anymore. For tests, it should be a development dependency.

**Comment by jdurand** [comment]
@dblock any timeline on merging this?

**Comment by dblock** [comment]
This PR won't be merged without the changes I'm asking for above.

**Comment by dblock** [comment]
Replaced with https://github.com/ruby-grape/grape/pull/1594.

**Comment by dblock** [comment]
Please see https://github.com/ruby-grape/grape/pull/1610 as a working end-to-end PR.

**Pull request closed without merging**